charmhelpers/contrib/openstack/utils.py

```diff
@@ -1450,20 +1450,33 @@ def some_hook(...):
 
     see core.utils.restart_on_change() for more details.
 
+    Note restart_map can be a callable, in which case, restart_map is only
+    evaluated at runtime.  This means that it is lazy and the underlying
+    function won't be called if the decorated function is never called.  Note,
+    retains backwards compatibility for passing a non-callable dictionary.
+
     @param f: the function to decorate
-    @param restart_map: the restart map {conf_file: [services]}
+    @param restart_map: (optionally callable, which then returns the
+        restart_map) the restart map {conf_file: [services]}
     @param stopstart: DEFAULT false; whether to stop, start or just restart
     @returns decorator to use a restart_on_change with pausability
     """
     def wrap(f):
+        # py27 compatible nonlocal variable.  When py3 only, replace with
+        # nonlocal keyword
+        __restart_map_cache = {'cache': None}
+
         @functools.wraps(f)
         def wrapped_f(*args, **kwargs):
             if is_unit_paused_set():
                 return f(*args, **kwargs)
+            if __restart_map_cache['cache'] is None:
+                __restart_map_cache['cache'] = restart_map() \
+                    if callable(restart_map) else restart_map
```

Comment:
Have we tested this during an OpenStack upgrade?

Reply:
If you think about how the code *used* to work, it was evaluated when the function was *defined*.  Now it just waits till the function is called.  If the issue is that the restart_map might change between when the function is defined (i.e. at load-up) vs when it is called (as part of the hook or a hook), then it might be a problem.  If not, then there is no issue.

If it's between invocations of the hooks then that is a fresh run of the python code, and it would be re-evaluated.

I've updated the .travis.yml as well.

Reply:
During an upgrade it is possible that restart_map will be *called* before and after. It also possible the return values will differ. I will need to test this.

Reply:
Okay, that seems fair enough.  We can remove the caching quite easily.  It might have been an optimisation too far! :)

Reply:
Agreed. And we can do that RSN. For now, I have tested it on a couple of charms and it does the right thing. I'll merge it.